### ISSUE TITLE
New filters for improved moderation

### ISSUE THREAD

**Issue by gsingh93** [kern/hackbot#9]
I have some proposed changes to improve the quality of posts in the main HH group. I'd be happy to implement some or even all of these myself. I wanted to post this issue to get approval and feedback before starting. Here are the changes:
1. Create a new user named Hackbot which will be used in the following items (is this against the Facebook TOS? Is there any alternative to this for the use cases below?).
2. Add an argument to the `/thread` command. The argument will be a set of predefined keywords explaining why the thread was closed. For example, `/thread editor` will close a thread comparing editors. Once a thread is closed, have Hackbot comment on the thread with a predefined message explaining why the thread is being closed. If the argument to the command doesn't exist, Hackbot won't post anything. This is so the mod who closed the thread can give a custom explanation if there is no predefined explanation. A detailed list of what should and shouldn't be allowed can be discussed later.
3. Add a `/redirect <group>` command. The `<group>` argument will be something like `python`, which in this case would cause Hackbot to comment with a link to the HH Python group. Whether this should be followed up with a `/thread` will be up to the moderator. Any user can run this command, not just mods.
4. Create a `/wiki` or `/faq`. Using this command will cause Hackbot to comment with a link to some page that answers the most common questions. Hopefully we can knock of all the "What are some good resources for X" questions with this command. Any user can run this command, not just mods.
5. Add a `/report` command. I think this will be easier than using the builtin FB functionality. Anyone can report a post. Reporting a post will send a notification to all of the mods. This could be useful as there aren't many mods so not all posts are seen, but any user can run `/report` on a thread that's becoming an editor war, which will notify the mods to close the thread. While I think this would be convenient, I can think of multiple reasons against it, so feel free to reject it.
6. Add a `/repost` command. Anyone can use this command. Running this command will make Hackbot post a comment with all of the previous reposts of the same link. Obviously it would be crazy to have Hackbot keep track of all the links posted ever, so this would be limited to some tweakable duration (i.e. all links older than 3 days would be removed from Hackbots cache).

**Comment by gsingh93** [comment]
If the new Hackbot user is against the TOS, the post could come from the account that issued the command. This shouldn't be a problem for the mods, as they can just give the app permission, and any users who wanted to use these tools could give the app permission themselves.

**Comment by gsingh93** [comment]
I'm also adding `/search <phrase>` to the list, which would post a link to searching a phrase in the group, which should handle the case when a question has been asked multiple times but hasn't yet been added to the FAQ.

@kern Can I get your opinion on these?

**Comment by kern** [comment]
Awesome feature suggestions @gsingh93. Let's knock these out one-by-one and create separate issues for the ones we want to implement:
1. Making a separate user for Hackbot is against Facebook's ToS, so currently Hackbot uses an access token with `user_groups` and `publish_actions` from a group admin.
2. Hackbot can't edit posts that it didn't make. This is rather unfortunate, but I also don't think a keyword is the best way to go about this. Perhaps instead we make a Chrome extension which adds a "Lock" button next to the Like and Comment buttons with a few prefilled phrases. Hackbot currently allows arbitrary text after the `/thread` command, so this could be done without any changes to Hackbot itself.
3. This `/redirect` functionality is identical to `/thread`, but uses better terminology for the situation. I'll make `/redirect` an alias of `/thread` for the time-being.
4. This is a _great_ idea. I'm thinking of `/star` or `/feature` as a command, which adds the post to a list on the site. The FAQ should be manually maintained, in my opinion, to make sure it's useful. This could be a jumping off point for an "HH Digest" or mailing list. I like it. :)
5. FB doesn't expose post reporting via the Graph API, and reports really shouldn't be public. Not sure how much more we can improve on reported posts given FB's limitations. I want the ability for users to report comments and provide a reason for their reporting, but I'm not sure the best way to go about it.

6 + `/search`. Again, would be cool to implement this, but it might be better as a Chrome extension.

**Comment by gsingh93** [comment]
Thanks for the reponse! I really like the `/star` idea, that would be really useful.

My second comment states that we could allow users to give Hackbot permission to post as them (like any other FB app). This satisfies the terms of service. Then anyone who cares enough to help out (and I'm sure there would be a lot of people who'd be willing to) could use the `/redirect` and other non-admin commands. The commands wouldn't work for anyone who didn't give Hackbot permission. What do you think about that?

The `/redirect` functionality I proposed wouldn't be a duplicate of `/thread` (as far as I understand it). The point of `/redirect` was essentially to say "here is a more appropriate group for this question", and not necessarily close the thread.

Any comments on the `/repost` command (given that you're OK with the idea of giving Hackbot permission to post as other users)?

**Comment by kern** [comment]
Ah, true. `/repost`, `/redirect`, `/search` and related commands could be implemented by deleting the user's original comment and having Hackbot comment an automated post on the user's behalf. Seems simple enough, and it might just work. :)

I've spun off each command into separate issues. `/repost` is going to take a little more time than `/redirect` or `/search` (both of which are really simple), but more importantly we're going to have to first allow users the ability to authenticate with Hackbot (see #6). I was planning on doing this after the rewrite I have going on the `sentiment` branch.

**Comment by gsingh93** [comment]
Awesome, sounds great.

If I get some time I'll try to knock out some of these commands myself. I'll let you know before I start working on them so we don't step on each others' toes.

**Comment by kern** [comment]
Sounds great, can't wait. :)
ᐧ
